### PR TITLE
Fix video decoder cache for multiple GPUs

### PR DESCRIPTION
**Pull request by JanuszL**
- the video decoder cache is global and doesn't
  account for multiple GPUs in the system. In the end,
  the decoder created for one GPU can be used in another
  causing failure. This PR adds device_id logic to the
  decoder cache

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

## Category:
**Bug fix** (*non-breaking change which fixes an issue*)

## Description:
- the video decoder cache is global and doesn't
  account for multiple GPUs in the system. In the end,
  the decoder created for one GPU can be used in another
  causing failure. This PR adds device_id logic to the
  decoder cache

### Affected modules and functionalities:
- frames decoder for the GPU

### Key points relevant for the review:
- performance

### Tests:

- [ ] Existing tests apply
- [x] New tests added
  - [x] Python tests
    - test_video.test_multi_gpu_video
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [x] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by JanuszL** [comment]
!build

File: dali/operators/reader/loader/video/frames_decoder_gpu.cc [hunk 1]
```diff
@@ -230,16 +235,22 @@ class NVDECCache {
...
+      for (auto &dec_cache_elm : dec_cache) {
+        for (auto &it : dec_cache_elm) {
+          cuvidDestroyDecoder(it.second.decoder);
+        }
       }
     }
-
-    std::unordered_multimap<cudaVideoCodec, DecInstance> dec_cache;
+    using codec_map = std::unordered_multimap<cudaVideoCodec, DecInstance>;
+    std::vector<codec_map> dec_cache;
```
Comment (mzient) [inline dali/operators/reader/loader/video/frames_decoder_gpu.cc:253]: I think we should have a mutex per device.

Comment (JanuszL) [reply dali/operators/reader/loader/video/frames_decoder_gpu.cc:255]: Fixed

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
!build

File: dali/operators/reader/loader/video/frames_decoder_gpu.cc [hunk 2]
```diff
@@ -218,8 +221,10 @@ class NVDECCache {
     }
 
     void ReturnDecoder(DecInstance &decoder) {
-      std::unique_lock lock(access_lock);
-      auto range = dec_cache.equal_range(decoder.codec_type);
+      int device_id = 0;
+      CUDA_CALL(cudaGetDevice(&device_id));
+        std::unique_lock lock(access_lock[device_id]);
```
Comment (mzient) [inline dali/operators/reader/loader/video/frames_decoder_gpu.cc:226]: Weird indent?

File: dali/operators/reader/loader/video/frames_decoder_gpu.cc [hunk 3]
```diff
@@ -230,18 +235,31 @@ class NVDECCache {
...
+      access_lock.resize(num_devices);
+    }
 
     ~NVDECCache() {
-      std::scoped_lock lock(access_lock);
-      for (auto &it : dec_cache) {
-        cuvidDestroyDecoder(it.second.decoder);
+      for (size_t i = 0; i < dec_cache.size(); ++i) {
+        auto &dec_cache_elm = dec_cache[i];
+        std::scoped_lock lock(access_lock[i]);
```
Comment (mzient) [inline dali/operators/reader/loader/video/frames_decoder_gpu.cc:248]: What do you need a `scoped_lock` for? I don't think it has any advantage over `lock_guard` when locking on e object.

File: dali/operators/reader/loader/video/frames_decoder_gpu.cc [hunk 4]
```diff
@@ -230,18 +235,31 @@ class NVDECCache {
...
       }
     }
-
-    std::unordered_multimap<cudaVideoCodec, DecInstance> dec_cache;
-
-    std::mutex access_lock;
+    using codec_map = std::unordered_multimap<cudaVideoCodec, DecInstance>;
+    std::vector<codec_map> dec_cache;
+
+    struct mutex_wrapper : std::mutex {
```
Comment (mzient) [inline dali/operators/reader/loader/video/frames_decoder_gpu.cc:257]: What is this wrapper for?

Comment (JanuszL) [reply dali/operators/reader/loader/video/frames_decoder_gpu.cc:257]: To create a vector of mutexes (raw mutex doesn't allow that). I'm open to other options.

Comment (mzient) [reply dali/operators/reader/loader/video/frames_decoder_gpu.cc:257]: The problem is `resize`, but you don't have to use it.
```
mutexes = std::vector<std::mutex>(32);
```

File: dali/operators/reader/loader/video/frames_decoder_gpu.cc (hunk 3 above)
Comment (JanuszL) [reply dali/operators/reader/loader/video/frames_decoder_gpu.cc]: Fixed

File: dali/operators/reader/loader/video/frames_decoder_gpu.cc (hunk 2 above)
Comment (JanuszL) [reply dali/operators/reader/loader/video/frames_decoder_gpu.cc]: Fixed

File: dali/operators/reader/loader/video/frames_decoder_gpu.cc (hunk 4 above)
Comment (JanuszL) [reply dali/operators/reader/loader/video/frames_decoder_gpu.cc]: Reworked

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
!build

**Review by szalpal — APPROVED** [review]

**Review by mzient — APPROVED** [review]

**Pull request merged**